RAG_supporters/dataset/torch_dataset.py

```diff
@@ -98,21 +145,276 @@ def __init__(
+
+            # Get cached LLM steering embedding
+            if (
+                self._llm_steering_embeddings_cache is not None
+                and idx in self._llm_steering_embeddings_cache
+            ):
+                return self._llm_steering_embeddings_cache[idx]
+
+            # Compute on-the-fly
+            if idx in self.llm_steering_texts and self.embedding_model:
+                steering_text = self.llm_steering_texts[idx]
+                emb = np.array(
+                    self.embedding_model.embed_query(steering_text), dtype=np.float32
+                )
+                return emb
+
+            if embedding_dim:
+                return np.zeros(embedding_dim, dtype=np.float32)
+            return None
+
+        elif self.steering_mode == SteeringMode.MIXED:
+            # Weighted combination of multiple steering embeddings
+            embeddings = []
+            weights = []
+
+            # Collect embeddings based on weights
+            if self.steering_weights.get("suggestion", 0) > 0 and suggestions:
+                if self._suggestion_embeddings_cache and suggestions[0] in self._suggestion_embeddings_cache:
+                    embeddings.append(self._suggestion_embeddings_cache[suggestions[0]])
+                    weights.append(self.steering_weights["suggestion"])
+
+            if self.steering_weights.get("cluster_descriptor", 0) > 0:
+                cluster_id = self._get_primary_cluster(idx)
+                if (
+                    cluster_id is not None
+                    and self._cluster_descriptor_embeddings_cache
+                    and cluster_id in self._cluster_descriptor_embeddings_cache
+                ):
+                    embeddings.append(
+                        self._cluster_descriptor_embeddings_cache[cluster_id]
+                    )
+                    weights.append(self.steering_weights["cluster_descriptor"])
+
+            if not embeddings:
+                if embedding_dim:
+                    return np.zeros(embedding_dim, dtype=np.float32)
+                return None
+
+            # Weighted average
+            weights = np.array(weights, dtype=np.float32)
+            weights = weights / weights.sum()  # Normalize
+            mixed_emb = sum(w * emb for w, emb in zip(weights, embeddings))
+            return mixed_emb.astype(np.float32)
+
+        return None
+
+    def _get_primary_cluster(self, idx: int) -> Optional[int]:
+        """Get primary cluster assignment for a sample."""
+        if not self.cluster_labels or idx not in self.cluster_labels:
+            return None
+
+        assignment = self.cluster_labels[idx]
+        if isinstance(assignment, int):
+            return assignment
+        elif isinstance(assignment, list) and assignment:
+            # For soft labels, return index with highest probability
+            return int(np.argmax(assignment))
+        return None
+
+    def _get_num_clusters(self) -> int:
```

Comment:
This should be done based on JSON file generated from `keyword_clustering.py`. (for example from clusters key)